single_kernel_mongo/events/tls.py

```diff
@@ -9,205 +9,225 @@
-        logger.debug("Request to set TLS private key received.")
-        if (
-            self.manager.state.is_role(MongoDBRoles.MONGOS)
-            and self.manager.state.config_server_name is None
-        ):
-            logger.info(
-                "mongos is not running (not integrated to config-server) deferring renewal of certificates."
+        for cert_requires in [self.peer_certificate, self.client_certificate]:
+            self.framework.observe(
+                cert_requires.on.certificate_available, self._on_certificate_available
             )
-            event.fail("Mongos cannot set TLS keys until integrated to config-server.")
-            return
-        if self.manager.state.upgrade_in_progress:
-            fail_action_with_error_log(
-                logger,
-                event,
-                "set-tls-private-key",
-                "Setting TLS keys during an upgrade is not supported.",
+
+        for relation_name in [
+            ExternalRequirerRelations.PEER_TLS.value,
+            ExternalRequirerRelations.CLIENT_TLS.value,
+        ]:
+            self.framework.observe(
+                self.charm.on[relation_name].relation_created,
+                self._on_tls_relation_created,
             )
-            return
-        try:
-            for internal in (True, False):
-                param = "internal-key" if internal else "external-key"
-                key = event.params.get(param, None)
-                csr = self.manager.generate_certificate_request(key, internal=internal)
-                self.certs_client.request_certificate_creation(certificate_signing_request=csr)
-                self.manager.set_waiting_for_cert_to_update(internal=internal, waiting=True)
-        except ValueError as e:
-            event.fail(str(e))
-
-    def _on_tls_relation_joined(self, event: RelationJoinedEvent) -> None:
-        """Handler for relation joined."""
-        if (
-            self.manager.state.is_role(MongoDBRoles.MONGOS)
-            and self.manager.state.config_server_name is None
-        ):
-            logger.info(
-                "mongos is not running (not integrated to config-server) deferring renewal of certificates."
+            self.framework.observe(
+                self.charm.on[relation_name].relation_broken,
+                self._on_tls_relation_broken,
             )
-            event.defer()
+
+        self.framework.observe(self.charm.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.charm.on.secret_changed, self._on_secret_changed)
+
+    @property
+    def tls_mapping(self) -> dict[bool, TLSCertificatesRequiresV4]:
+        """Mapping of boolean to a TLS requirer instance.
+
+        The boolean value is True if the requirer is for internal certificates,
+        and False for the client certificates.
+        """
+        return {True: self.peer_certificate, False: self.client_certificate}
```

Comment:
This is an odd data structure, why not a dict[str, ..] ?

Reply:
Because we always identify the internal/external by a boolean, and this structure (even if odd) was the easiest for me if I wanted a dict-like access.
However I could bring that to a different shape